### DISCUSSION TITLE
[Unknown Mod][Spells] Tweaks to "View Mines", "View Resources", "View Artifacts", "View Heroes", "View Towns" and "View All" spells [WIP]

### DISCUSSION THREAD

**Discussion by LeHerosInconnu** [ihhub/fheroes2#2544] [Ideas]
I would like to make these spells to remain active all day long just like the "Identify Hero" spell.
As on a large map with many heroes, it's quite annoying to have to cast the same spells several times per turn and it's not really user friendly.

This will be an option:
"world: view spells are active for the entire duration of the turn"

So the player casts the spell only once a day (you cast a spell to enter the spell screen, look at the spell screen and then close the spell screen) and when you open the View World screen, all the information provided by the spell are displayed for the whole day, no need to cast the spell several times.

The player can cast several spells (e.g. View Resources, View Heroes, etc.), all the information they provide are displayed at the same time in the View World screen.
![View World H2 all view spells avec boutons 01](https://user-images.githubusercontent.com/43583994/103768405-baeaeb00-5022-11eb-9c28-8c1bf76cd4de.png)

When at least one of the "View" spells is active, buttons are added (in the example to the right of the "Zoom" button, with active "View Mines" spell) to allow the player to access the spell screen.
When the mouse cursor is positioned over the "View panel" area, the player can use the mouse wheel to scroll through the different screens, or use the right and left arrow keys on the keyboard for the same result.
When the mouse cursor is positioned over the general view area or the minimap area, the player can use the mouse wheel to zoom in and out (same action as the "Magnifying glass" button), or use the up and down arrow keys on the keyboard for the same result.
![View World H2 mine and boutons 01](https://user-images.githubusercontent.com/43583994/103768525-f4235b00-5022-11eb-8e9d-903fc14a0b29.png)

In each "View" spell screen, a checkbox with the text "Auto" is added (in the example to the left of the "Zoom" button) to allow the player to select the automatic casting of the spell at the beginning of each turn.
![View spells H2 boutons et auto 001](https://user-images.githubusercontent.com/43583994/103768658-3187e880-5023-11eb-805b-63ad2de20031.png)

When automatic casting is activated, the same hero who cast the spell initially will cast the spell automatically at the beginning of each turn.
If the hero does not have enough spell points to cast a spell, a message is displayed in a window at the beginning of the turn telling the player that the spell cannot be cast, and the spell "auto checkbox" is unchecked automatically.
If more than one spell casting message is to be displayed, they are all displayed in the same window (the text is for explanation only, it must be adapted for all possibilities).
![Heroes 2 Screen 56](https://user-images.githubusercontent.com/43583994/103768817-7dd32880-5023-11eb-8506-be03d220f405.png)

If the hero has to cast several spells but does not have enough spell points to cast all the spells, the spells are cast in the order in which they were originally cast as long as it is possible to cast them.

Examples:

You cast View Mines.
![Heroes 2 Screen 53a01](https://user-images.githubusercontent.com/43583994/103768974-c7bc0e80-5023-11eb-9510-9514879b10a7.png)

1) The "Mine symbol" surrounded by a green frame shows that the spell is active for the day. Green colored frame, the spell is active. Red colored frame, the spell is not active.
The "= mine" green colored text shows that the mines are displayed in the informations area map. Green colored text, the mines are displayed. Red colored text, the mines are not displayed.
You can click on the "= mine" colored text to show/hide the mines in the informations area map.

2) The "Resource symbol" surrounded by a green frame shows that the mines of this resource are displayed in the informations area map. Green colored frame, the mines are displayed. Red colored frame, the mines are not displayed.
You can click on each "Resource symbol" to show/hide the mines of each resource in the informations area map.
The "all" button allows to show/hide all the mines at the same time.
Click the "all" green colored button, all the mines are hidden, the button becomes red colored.
Click the "all" red colored button, all the mines are shown, the button becomes green colored.

E.g. All the mines are displayed. You click the "ore resource" button to hide the mines of this resource on the map. The "ore resource" button becomes red colored, the "all" button becomes yellow colored, the "= mine" text becomes yellow colored (the yellow color shows that some mines are not displayed).
Click the "all" yellow colored button, all mines are shown, the "all" button becomes green colored.
Click the "= mine" yellow colored text, all mines are shown, the "= mine" text becomes green colored.

You cast View Mines, then exit the View Mines screen. You open the View World screen the same day. The spell is still active, the informations are displayed.
![Heroes 2 Screen 53a02](https://user-images.githubusercontent.com/43583994/103769090-fb973400-5023-11eb-9187-9cdcae5ea62f.png)

1) The "Mine symbol" surrounded by a green frame shows that the spell is active.
You can click on the "= mine" green colored text to hide the mines in the informations area map.
You can click on the "Mine symbol" green colored to access the "View Mines" settings (show/hide each type of mine). Another click on the "Mine symbol" closes the "View Mines" settings.

You cast View Resources.
![Heroes 2 Screen 53a03](https://user-images.githubusercontent.com/43583994/103769151-0baf1380-5024-11eb-9c1c-6a46320735a4.png)

1) You can click the "= wood", "= gems", etc. colored text to show/hide the resources in the informations area map.

You cast View Resources.
![Heroes 2 Screen 53a04](https://user-images.githubusercontent.com/43583994/103769363-647eac00-5024-11eb-83a8-3c4886c6a40b.png)

1) You click the "= gold" green colored text to hide the gold resources on the map, the "= gold" text becomes red colored, the "all" button becomes yellow colored (the yellow color shows that some resources are not displayed).

2) The "= resource" text becomes yellow colored (the yellow color shows that some resources are not displayed).

You cast View Resources, then exit the View Resources screen. You open the View World screen the same day. The spell is still active, the informations are displayed.
![Heroes 2 Screen 53a05](https://user-images.githubusercontent.com/43583994/103769384-6ea0aa80-5024-11eb-9ba6-ce63637d6de9.png)

1) The "Resource symbol" surrounded by a green frame shows that the spell is active. 
The "= resource" yellow colored text shows that some resources are not displayed.
You can click the "Resource symbol" green colored to access the "View Resources" settings (show/hide each type of resource). Another click on the "Resource symbol" closes the "View Resources" settings.
You can click the "= resource" yellow colored text, all resources are shown, the "= resource" text becomes green colored.

You cast View Mines, then exit the View Mines screen. You cast View Resources, then exit the View Resources screen. You open the View World screen the same day. The spells are still active, the informations are displayed.
![Heroes 2 Screen 53a06](https://user-images.githubusercontent.com/43583994/103769412-7d875d00-5024-11eb-8202-2357e5782964.png)

1) The "Mine symbol" and the "Resource symbol" surrounded by green frames shows that the spells are active.
The "= mine" green colored text shows that the mines are displayed in the informations area map.
The "= resource" yellow colored text shows that some resources are not displayed in the informations area map.

With the "View All" spell, the "Town symbol" (flag), "Mine symbol", "Hero symbol", "Artifact symbol" and "Resource symbol" are all surrounded by a green frame.
You can access all the settings (if any). The program remembers all your last settings, so no need to set them each day if you cast the spells.
While the "View All" spell is active, when you access the settings screen for the others "View" spells, the checkbox with the text "Auto" are not displayed, as you access the screens to change the settings and not the "real" "View" spell screens.

If the player tries to cast the View Mines spell while the spell is already active, a window is displayed (right in the Magic Book window) to inform the player.
![Heroes 2 Screen 53a07](https://user-images.githubusercontent.com/43583994/103769446-909a2d00-5024-11eb-8a28-b55f3647fdad.png)

The "View World" button is accessible in this window to allow the player to go directly to the "View World" screen.

If the player tries to cast the View Mines spell (or others) while the View All spell is already active, a window is displayed (right in the Magic Book window) to inform the player this is not necessary.
![Heroes 2 Screen 53a08](https://user-images.githubusercontent.com/43583994/103769463-9b54c200-5024-11eb-9f16-cedecbe29cb6.png)

The "View World" button is accessible in this window to allow the player to go directly to the "View World" screen.

You open the View World screen.
![Heroes 2 Screen 53a09](https://user-images.githubusercontent.com/43583994/103769487-a9a2de00-5024-11eb-9eb1-e65091a86e8b.png)

1) "= mine" in orange colored text indicates that at least one of your heroes (or captains) can cast the View Mines spell. You can click on the orange colored text "= mine" to cast the spell.

You click on the orange colored text "= mine". The "Cast View Mines panel" is displayed (replaces the "View World panel").
![Heroes 2 Screen 53a10b03b](https://user-images.githubusercontent.com/43583994/103769514-b7f0fa00-5024-11eb-9f72-52afa2a24715.png)

1) All the heroes/captains who can cast the spell are displayed. They are classified with respect to the remaining spell points, from biggest to smallest.
Note: The captain who can cast some adventure spells is another option that will be described in another post.

To select the caster, simply click on his portrait (Ranloo is selected in this example). By default, the first hero is selected (the one with the biggest number of spell points left).
A button to display only heroes or only captains or both (small hero portrait graphic) is also available in this screen.

To cast the spell, click on the "Cast Spell" button (hand), once your choice is made.

A checkbox allows to automatically cast the spell at the beginning of the day. Checked, the selected hero casts the spell daily. Unchecked, the selected hero casts the spell one time (for the current day).
The checkbox in the screen of the "View" spell that is cast is also updated.

The player can click the EXIT button to exit the panel without casting the spell and go back to the "View World panel".

After the spell is casted, the "Cast view Mines panel" is closed and the player go to the "View Mines" screen.

In the "View World panel" and "View Mines panel", the "Mine symbol" becomes surrounded by a green frame and the "= mine" text becomes green colored.

The icons of the different spells are also directly displayed over the Fog of War on the adventure map.
This could be a sub-option of the general "View" spells option:
- "world: view spells are active for the entire duration of the turn"
- "world: view spells icons are displayed over the Fog of War"

![Heroes 2 Screen 53a11](https://user-images.githubusercontent.com/43583994/103769570-d0611480-5024-11eb-8b4f-68f6592ea318.png)

1) On the adventure map, the mines symbols are displayed over the hidden territory.
This applies to all spells, View Resources, View Artifacts, etc.

All settings for displaying different icons in the "View World" and "View" spells screens are also applied for icons displayed over the Fog of War on the adventure map.
![View spells H2 view mines view resources 009](https://user-images.githubusercontent.com/43583994/103769592-da831300-5024-11eb-8ae1-bf80842eb27e.png)
![View spells H2 view artifacts view towns 02 view heroes 02 010](https://user-images.githubusercontent.com/43583994/103769639-ebcc1f80-5024-11eb-89b2-4330112e4baf.png)

When the player casts the "View All" spell and exits the "View All" spell screen, the Fog of War is removed from the adventure map.
The Fog of War from the unrevealed squares is no longer displayed as long as the spell remains active.
The symbols displayed over the Fog of War are no longer displayed.

When the "View All" spell is active, on the adventure screen, the player can press the "F1" key to alternatively display/hide the Fog of War (one press on the "F1" key, the Fog of War is displayed, another press on the "F1" key the Fog of War is no longer displayed).
The symbols are also displayed over the hidden territory when the Fog of War is displayed.
The player can also press the central mouse button to obtain the same effect.

When the player casts the "View All" spell and one or more "View" spells are already active and set to "Auto" casting, a message window informs the player of this and a question is asked if he wishes to cancel the automatic casting of these spells (this window is displayed right after the View All screen is open).
A checkbox with "Auto" is also available in this window to allow the automatic casting of the "View All" spell.

In the hero screen, the graphic with a "3" over the portrait of the hero indicates the number of adventure spells currently being automatically cast each day by the hero.
![View spell actif hero screen 02b](https://user-images.githubusercontent.com/43583994/103769704-11f1bf80-5025-11eb-8787-03a86b824f01.png)

Right-clicking on this graphic displays a temporary information window that shows the spells automatically cast each day by the hero.
A left-click on this graphic displays an information window with an "OKAY" button that indicates the spells automatically cast each day by the hero.
It is possible to uncheck the box for the automatic casting of each of the spells cast by this hero in this window.

In the Magic Book, the graphic with a white cross next to the spell sprite indicates that the spell is already cast for the current turn (by the hero or by any hero (or captain) of the player) and that the spell is cast automatically at the beginning of each turn.
Right-clicking on this graphic displays a temporary information window that indicates which hero is casting the spell automatically.
A left click on this graphic displays an information window with an "OKAY" button that indicates which hero is casting the spell automatically.
It is possible to uncheck the box for the automatic casting of the spell by the hero in this window.

In the Magic Book, the graphic with a white square next to the spell sprite indicates that the spell is already cast for the current turn.
Right-clicking on this graphic displays a temporary information window that shows which hero has cast the spell for the current round.
A left-click on this graphic displays an information window with an "OKAY" button that indicates which hero has cast the spell for the current turn.

In the hero screen, the white square in the graphic with a "3" over the portrait of the hero indicates that at least one adventure spell has currently been cast by the hero for the current day only.
![View spell actif hero screen 02b2](https://user-images.githubusercontent.com/43583994/103770303-0e126d00-5026-11eb-8dd3-c08cb7d2c61c.png)

On the small portrait of the hero, a mark (white cross) is also added to indicate to the player that the hero casts a spell automatically every day (Ranloo, second portrait on the right in this example).
![Heroes 2 Screen 53a10b03b](https://user-images.githubusercontent.com/43583994/103769514-b7f0fa00-5024-11eb-9f72-52afa2a24715.png)

On Jojosh's portrait (third portrait on the right), the mark (white square) indicates that he has cast at least one "View" spell for the current turn (but not automatically).
On Atlas's portrait (fourth portrait on the right), the mark (white square and white cross) indicates that he has cast at least one "View" spell for the current turn (spell cast for the current turn but not automatically) and that he casts at least one "View" spell automatically every day.
![Heroes 2 Screen 53a10b06](https://user-images.githubusercontent.com/43583994/103770475-5c277080-5026-11eb-937a-6a2ec3097d6a.png)

In the View World screen, a "Cast Spell" button (hand button at left the "Zoom" button in the example) is also added to allow the player to make his heroes cast adventure spells.
This button is only displayed if at least one adventure spell is known by any of his heroes (or captains).
![Heroes 2 Screen 53a02b2](https://user-images.githubusercontent.com/43583994/103770586-8c6f0f00-5026-11eb-986c-a1977c2eee63.png)

Once the button is pressed, a Magic Book is displayed with all the "View" spells known by any of his heroes (or captains).
Note: This is not the specific book of a hero, but a simple and already known way to select one of the available spells.
The number of spell points is therefore not displayed.
![Heroes 2 Screen 53a02c04](https://user-images.githubusercontent.com/43583994/103770597-91cc5980-5026-11eb-9cef-d5f3030db2a3.png)

Once a left click on the spell is done, the "Cast View spell panel" is displayed with the name of the spell chosen in the title and the player can select one of the heroes (or captains) who knows the spell to cast it.
![Heroes 2 Screen 53a10b06](https://user-images.githubusercontent.com/43583994/103770475-5c277080-5026-11eb-937a-6a2ec3097d6a.png)

When the player casts a "View" spell that is already active, a message window informs the player about it and a question is asked if he wants to cast the spell with the new hero from the next turn in place of the current hero.
A checkbox with "Auto" is also available in this window to allow the automatic casting of the "View" spell.

When the player casts a "View" spell and the "View All" spell is already active, a message window informs the player about it and a question is asked if he wants to cast the spell with the new hero from the next turn in place of the current hero and cancel the casting of the "View All" spell at the same time.
A checkbox with "Auto" is also available in this window to allow the automatic casting of the "View" spell.
 
When the player casts the "View All" spell and the "View All" spell is already active, a message window informs the player about it and a question is asked if he wants to cast the spell with the new hero from the next turn in place of the current hero.
A checkbox with "Auto" is also available in this window to allow the automatic casting of the "View All" spell.

When the player casts the "View All" spell and one or more "View" spells are already active, a message window informs the player about it and a question is asked if he wants to cast this spell with the new hero from the next turn in place of the current hero(es) and cancel the casting of the "View" spell(s) at the same time.
A checkbox with "Auto" is also available in this window to allow the automatic casting of the "View All" spell.

In the "Wiew World" screen or in the screen of one of the "View" spells, when the player moves the mouse cursor over the main area or over the minimap a crosshair is displayed in both areas, and when the player left clicks on the main area or on the minimap, the screen is closed and the view of the adventure map is directly positioned at this previously selected location.
![View World H2 all view spells avec boutons et curseurs et reticules flèche 01](https://user-images.githubusercontent.com/43583994/103770679-b294af00-5026-11eb-949d-754e2ac3d3e9.png)

In all screens:
Right-click information windows are available for buttons, texts and symbols.
Right clic info window on "Auto" and checkbox also displays the name of the hero casting the spell and his remaining and total spell points.

Images are for explanation only (and some screens are missing).
Screen and window texts need to be reworked with titles and images added, etc. :)

**Edit 2021/01/07.**

Related post: https://github.com/ihhub/fheroes2/discussions/2549.

**Comment by vincent-grosbois** [comment]
I skipped over some parts of the post, but remember that "View All" does'nt show things like pick-up objects (chests etc)
So if you remove the fog of war when you cast "view all", either it will show the chests, or not show the chests but be super confusing

**Reply by LeHerosInconnu** [reply]
@vincent-grosbois.

> I skipped over some parts of the post, but remember that "View All" does'nt show things like pick-up objects (chests etc)
> So if you remove the fog of war when you cast "view all", either it will show the chests, or not show the chests but be super confusing

In the original game, the "View All" spell shows absolutely everything.
So it's just its normal operation applied to the adventure map.

In original game:

![View All H2 02](https://user-images.githubusercontent.com/43583994/103784816-b0d4e680-503a-11eb-99eb-05af6f1baeee.png)

**Comment by vincent-grosbois** [comment]
So it shows the sprites for the chests, artifacts etc  ? Is that the same for "view world" ? 

edit: what about monsters?

**Reply by LeHerosInconnu** [reply]
@vincent-grosbois.

> So it shows the sprites for the chests, artifacts etc ? Is that the same for "view world" ?

As the View World screen will show all active "View" spells, everything will be visible when the "View All" spell is active.

And this is already the case in the original game.

In original game:

![View World H2 02](https://user-images.githubusercontent.com/43583994/103786556-d95de000-503c-11eb-8efa-e3f7b99ddde2.png)

> edit: what about monsters?

Everything is everything. :)
You can see Green Dragons near the mouse cursor.

**Comment by minuspinus** [comment]
it's easier to make the cost of re-casting 0 mana for one day
